libs/gltfio/src/MorphHelper.cpp

```diff
@@ -0,0 +1,366 @@
+    const cgltf_mesh* mesh = node->mesh;
+    const cgltf_primitive* prims = mesh->primitives;
+    const cgltf_size prim_count = mesh->primitives_count;
+
+    // If the original VertexBuffer can be used, just use that instead of creating a new one.
+    // In such a case, the MorphHelper is not the owner of the vertex buffer.
+    const ubyte4 kOriginalPrimaryIndices(0, 1, 2, 3);
+    result.owner = !isSubsetOf(primaryIndices, kOriginalPrimaryIndices);
+    if (!result.owner) {
+        result.primaryIndices = kOriginalPrimaryIndices;
+    }
+
+    for (cgltf_size pi = 0; pi < prim_count; ++pi) {
+        const cgltf_primitive& prim = prims[pi];
+        RenderableManager::PrimitiveType prim_type;
+        getPrimitiveType(prim.type, &prim_type);
+        const auto& gltfioPrim = mAsset->mMeshCache.at(mesh)[pi];
+        VertexBuffer* const vertices = result.owner ?
+                createVertexBuffer(prim, gltfioPrim.uvmap, primaryIndices) :
+                gltfioPrim.vertices;
+        result.primitives.push_back({
+            .vertices = vertices,
+            .indices = gltfioPrim.indices,
+            .type = prim_type,
+        });
+    }
+    mMorphTable[entity].permutations.emplace_back(std::move(result));
+}
+
+// Creates a VertexBuffer for a given permutation of "primary indices" (i.e. set of 4 targets). In
+// some ways this mimics AssetLoader::createPrimitive() but it is simpler and less efficient because
+// it immediately clones and uploads buffer data.
+//
+// TODO: This strategy is inefficient because it results in duplication of data. For example, an
+// unmorphed UV0 attribute will have duplicated uploads and duplicated GPU data. We will soon be
+// adding a new API to Filament to fix this.
+VertexBuffer* MorphHelper::createVertexBuffer(const cgltf_primitive& prim, const UvMap& uvmap,
+        ubyte4 primaryIndices) {
+
+    // Determine the number of vertices by looking at the first usable attribute.
+    const uint32_t vertexCount = [&prim]() {
+        uint32_t count = 0;
+        for (cgltf_size aindex = 0; aindex < prim.attributes_count; aindex++) {
+            const cgltf_attribute& attribute = prim.attributes[aindex];
+            const cgltf_accessor* accessor = attribute.data;
+            if (accessor) {
+                count = accessor->count;
+                break;
+            }
+        }
+        return count;
+    }();
+
+    // This creates a copy because we don't know when the user will free the cgltf source data.
+    // For non-morphed vertex buffers, we use a sharing mechanism to prevent copies, but here
+    // we just want to keep it as simple as possible.
+    auto createBufferDescriptor = [](const cgltf_accessor* accessor) {
+        auto bufferData = (const uint8_t*) accessor->buffer_view->buffer->data;
+        const uint8_t* data = computeBindingOffset(accessor) + bufferData;
+        const uint32_t size = computeBindingSize(accessor);
+        uint8_t* clone = (uint8_t*) malloc(size);
```

Comment:
this makes me very nervous, the `malloc` happens at a different place frmo the `free`. Could this be refactored to ahve the malloc/free in the same object, or, use a `unique_ptr`?

Reply:
The `free` is actually just two lines after the malloc but you may've missed it because it was wrapped.  I fixed this to be more obvious.  (emscripten does not let me pass `free` directly into the callback)

Reply:
ah yes... duh.